byron/ledger/executable-spec/test/Ledger/Relation/Properties.hs

```diff
@@ -0,0 +1,223 @@
+  -> Set (Domain r)
+  -> r
+  -> m ()
+propDomainExclusionAndSetDifference find x s r
+  = find x ((dom r `Set.difference` s) ◁ r)
+    ===
+    find x (s ⋪ r)
+
+-- | (dom r1 ∪ s) ⋪ (r1 ∪ r2) == (dom r1 ∪ s) ⋪ r2
+--
+--  Excluding 'dom r1 ∪ s' from the domain of 'r1 ∪ r2'
+--  is the same as excluding 'dom r1 ∪ s' from the domain of just r2
+--  (since 'dom r1 ⋪ r1' gives an empty relation)
+propDomainExclusionAndUnion
+  :: (MonadTest m, Relation r, Ord (Domain r), Ord (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> r
+  -> m ()
+propDomainExclusionAndUnion find x s r1 r2
+  = find x ((dom r1 `Set.union` s) ⋪ (r1 ∪ r2))
+    ===
+    find x ((dom r1 `Set.union` s) ⋪ r2)
+
+
+-- | s ◁ (r1 ∪ r2) == (s ◁ r1) ∪ (s ◁ r2)
+--
+-- Restricting the domain of 'r1 ∪ r2' to 's' is the same as restricting the
+-- domains of r1 and r2 seperately and then taking the relation union of that.
+{- TODO @uroboros keep this?
+propDomainRestrictionAndUnion
+  :: (MonadTest m, Relation r, Ord (Domain r), Ord (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> r
+  -> m ()
+propDomainRestrictionAndUnion find x s r1 r2
+  = find x (s ◁ (r1 ∪ r2))
+    ===
+    find x ((s ◁ r1) ∪ (s ◁ r2))
+-}
+
+--------------------------------------------------------------------------------
+-- Functions abstracted over 'Relation r'
+--------------------------------------------------------------------------------
+
+-- | Combined property abstracted over 'Relation r'
+propRelation_
+  :: ( Relation r, Ord (Domain r), Ord (Range r)
+     , Show r, Show (Domain r), Show (Range r))
+  => Gen r
+  -> Gen [Domain r]
+  -> (Domain r -> r -> Maybe (Range r))
+  -> PropertyT IO ()
+propRelation_ genRelation_ genList_ find_
+  = do
+      r1 <- forAll $ genRelation_
+      s <- forAll $ Set.fromList <$> genList_
+      -- pick an element from 's ∪ dom r'
+      x <- forAll $ Gen.element (Set.toList $ s `Set.union` dom r1)
+
+      classify "(dom r) ∩ s == dom (s ◁ r)" $
+               Set.member x (dom r1 `Set.intersection` s)
+      propDomainRestrictionAndIntersection x s r1
+
+      classify "(dom r ∩ s) ◁ r == s ◁ r" $
+               Set.member x (dom ((dom r1 `Set.intersection` s) ◁ r1))
+      propDomainRestrictionAndIntersectionB find_ x s r1
+
+      classify "(dom r \\ s) ◁ r == s ⋪ r" $
+               Set.member x (dom ((dom r1 `Set.difference` s) ◁ r1))
+      propDomainExclusionAndSetDifference find_ x s r1
+
+      r2 <- forAll $ genRelation_
+      -- pick an element from 's ∪ dom r1 ∪ dom r2'
+      x' <- forAll $ Gen.element (Set.toList $ s `Set.union` dom r1 `Set.union` dom r2)
+
+      classify "(dom r1 ∪ s) ⋪ (r1 ∪ r2) == (dom r1 ∪ s) ⋪ r2" $
+               Set.member x' (dom ((dom r1 `Set.union` s) ⋪ (r1 ∪ r2)))
+      propDomainExclusionAndUnion find_ x' s r1 r2
+
+      {- TODO @uroboros fails for Bimap, but could make it pass with
```

Comment:
Oh, interesting. Is this a problem with the property or with the `Bimap` instance?

Reply:
the problem is that for BiMap in general this property does not hold (but does for Map and Set relation instances)

s ◁ (r1 ∪ r2) == (s ◁ r1) ∪ (s ◁ r2)

For Bimap it fails because the Union of Bimap relations does not allow for duplicates in the Range, e.g. 
r1 = {0 -> 1} , r2 = {2 -> 1} , s = {0} 
LHS = {}
RHS = {0 -> 1}

Which shows how much more restrictive the Bimap relation is than for Map and Set. 
At first I thought that I'd test with more restricted Bimaps (only uninteresting ones though!), and preserve the prop for Map and Set, but I'll just lose it altogether and try find a weaker property that works for all relations